odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+from numbers import Number
+
+from odl.operator.operator import (
+    Operator, OperatorComp, OperatorLeftScalarMult, OperatorRightScalarMult,
+    OperatorRightVectorMult, OperatorSum)
+from odl.operator.default_ops import (ResidualOperator, IdentityOperator,
+                                      ConstantOperator, ScalingOperator)
+from odl.set.space import LinearSpaceVector
+from odl.solvers.advanced import (proximal_arg_scaling, proximal_translation,
+                                  proximal_quadratic_perturbation)
+
+
+__all__ = ('Functional', 'FunctionalLeftScalarMult',
+           'FunctionalRightScalarMult', 'FunctionalComp',
+           'FunctionalRightVectorMult', 'FunctionalSum', 'FunctionalScalarSum',
+           'TranslatedFunctional', 'ConvexConjugateTranslation',
+           'ConvexConjugateFuncScaling', 'ConvexConjugateArgScaling',
+           'ConvexConjugateLinearPerturb')
+
+
+class Functional(Operator):
+
+    """Implementation of a functional class.
+
+    A functional is an operator ``f`` that maps from some domain ``X`` to the
+    field of scalars ``F`` associated with the domain:
+
+        ``f : X -> F``.
+
+    Notes
+    -----
+    The implementation of the functional class assumes that the domain
+    :math:`X` is a Hilbert space and that the field of scalars :math:`F` is a
+    is the real numbers. It is possible to create functions that do not fulfil
+    these assumptions, however some mathematical results might not be valide in
+    this case. For more information, see
+    http://odl.readthedocs.io/guide/in_depth/functional_guide.html.
+    """
+
+    def __init__(self, domain, linear=False, grad_lipschitz=np.inf):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain : `Set`
```

Comment:
Also here: consider `space` instead of `domain`. Most people are more familiar with the "space" notion, and here there is no ambiguity.